Source/Diagnostics/ReducedDiags/FieldReduced.cpp

```diff
@@ -0,0 +1,258 @@
+                    ofs << "max_Bz_lev"+std::to_string(lev)+" (T)";
+                    ofs << m_sep;
+                    ofs << "[" + std::to_string(m_offset_maxField+10+8*lev) + "]";
+                    ofs << "max_|B|_lev"+std::to_string(lev)+" (T)";
+                }
+            }
+
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+
+}
+// end constructor
+
+// function that computes field energy and/or maximum field values
+void FieldReduced::ComputeDiags (int step)
+{
+
+    // Judge if the diags should be done
+    if ( (step+1) % m_freq != 0 ) { return; }
+
+    // get WarpX class object
+    auto & warpx = WarpX::GetInstance();
+
+    // get number of level
+    const auto nLevel = warpx.finestLevel() + 1;
+
+    // loop over refinement levels
+    for (int lev = 0; lev < nLevel; ++lev)
+    {
+
+        // get MultiFab data at lev
+        const MultiFab & Ex = warpx.getEfield(lev,0);
+        const MultiFab & Ey = warpx.getEfield(lev,1);
+        const MultiFab & Ez = warpx.getEfield(lev,2);
+        const MultiFab & Bx = warpx.getBfield(lev,0);
+        const MultiFab & By = warpx.getBfield(lev,1);
+        const MultiFab & Bz = warpx.getBfield(lev,2);
+
+        if (m_fieldEnergy)
+        {
+            // get cell size
+            Geometry const & geom = warpx.Geom(lev);
+            auto domain_box = geom.Domain();
+#if (AMREX_SPACEDIM == 2)
+            auto dV = geom.CellSize(0) * geom.CellSize(1);
+#elif (AMREX_SPACEDIM == 3)
+            auto dV = geom.CellSize(0) * geom.CellSize(1) * geom.CellSize(2);
+#endif
+
+            // compute E squared
+            Real tmpx = Ex.norm2(0,geom.periodicity());
+            Real tmpy = Ey.norm2(0,geom.periodicity());
+            Real tmpz = Ez.norm2(0,geom.periodicity());
+            Real Es = tmpx*tmpx + tmpy*tmpy + tmpz*tmpz;
+
+            // compute B squared
+            tmpx = Bx.norm2(0,geom.periodicity());
+            tmpy = By.norm2(0,geom.periodicity());
+            tmpz = Bz.norm2(0,geom.periodicity());
+            Real Bs = tmpx*tmpx + tmpy*tmpy + tmpz*tmpz;
+
+            // save data
+            m_data[lev*3+1] = 0.5 * Es * PhysConst::ep0 * dV;
+            m_data[lev*3+2] = 0.5 * Bs / PhysConst::mu0 * dV;
+            m_data[lev*3+0] = m_data[lev*3+1] + m_data[lev*3+2];
+        }
+
+        if (m_maxField)
+        {
+            // get Maximums of E field components
+            m_data[m_offset_maxField+lev*8] = Ex.norm0();
+            m_data[m_offset_maxField+lev*8+1] = Ey.norm0();
+            m_data[m_offset_maxField+lev*8+2] = Ez.norm0();
+
+            // get Maximums of B field components
+            m_data[m_offset_maxField+lev*8+4] = Bx.norm0();
+            m_data[m_offset_maxField+lev*8+5] = By.norm0();
+            m_data[m_offset_maxField+lev*8+6] = Bz.norm0();
+
+            // Create temporary MultiFAB to be filled with |E| and |B| squared
```

Comment:
Hm, do we really need that intermediate MultiFAB or can we just reduce this on the fly?

Reply:
It's probably not necessary but it's the first simple solution that I found. :) It's not optimal at all, especially in terms of memory usage but I figured that we must have some spare memory for temporary MultiFABs in other parts of the code so it shouldn't be a problem anyway.

If we want something more efficient, maybe we can use `amrex::ReduceMax` provided that it works with MultiFABs that have different index types. Do you see another solution?

Reply:
I think there should be something, let me cc @atmyers for a suggestion.

Reply:
Pinging @ax3l and @atmyers again on this. (we're already using this diag in production runs so it would make things easier for us if it was merged into development)

@atmyers If you don't have time to go through the code, the main idea is that we want to compute the maximum value of `Ex**2 + Ey**2 + Ez**2`. So far we create a temporary `MultiFAB`, manually fill it in an `MFIter` loop with `arrtemp(i,j,k) = arrEx(i,j,k)*arrEx(i,j,k) + arrEy(i,j,k)*arrEy(i,j,k) + arrEz(i,j,k)*arrEz(i,j,k);` and then use `MultiFAB::max()`. Axel suggests that there should be a better way of doing this which does not involve an intermediate `MultiFAB`. Do you see how we could do this? (noting that `Ex`, `Ey` and `Ez` have in general a different staggering, which makes the result slightly inaccurate, but we're fine with this)

Reply:
You could express this operation using the reduction method here:

https://github.com/AMReX-Codes/amrex/blob/development/Tutorials/GPU/ParallelReduce/main.cpp#L212

The lambda function could capture multiple `Array4`s of arbitrary `IndexType` and perform both the interpolation to cell centers and the computation of E^2 for each cell.

Reply:
Thanks a lot, I've used that in the last commit.

I didn't do the interpolation to cell centers in the lambda because the functions that do interpolations in WarpX update an `Array4` rather than return an interpolated value, so it's not very convenient to use them here. But I will keep this possibility in mind for the future.